Markdown + RSS Parser for .NET Standard, MarkdownTextBlock Style Improvements, Syntax Highlighting

WilliamABradley opened the pull request:
- [x] Moved Markdown Parser (along with RSS Parser) to new project `Microsoft.Toolkit.Parsers`.

- [x] Made required Parsing Classes and Enums public (I know I could have done InternalsVisibleTo, but then the Parser wouldn't be extensible.

- [x] Split the Markdown Parser from MarkdownTextBlock.

- [x] Create ~~IMarkdownRenderer~~ MarkdownRendererBase, make XamlRenderer inherit it.

- [x] Replace JavaScript based Syntax Highlighter in Sample App with ColorCode-Universal.

- [x] Add CodeBlock addornments and DocFX formatting in Sample App.

- [x] Add Parsing of Html \<a> tags into new LinkAnchorInline.

- [x] Fix rendering of Code Blocks nested in Lists.

I also did some Restyling of the default Markdown to be further in line with Reddit and GitHub.

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[x] Code style update (formatting)
[x] Refactoring (no functional changes, no api changes)
[ ] Build or CI related changes
[x] Documentation content changes
[x] Sample app changes
[ ] Other... Please describe:
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Sample in sample app has been added / updated (for bug fixes / features)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->

- Namespaces for Parser elements change from Microsoft.Toolkit.Uwp.UI.Controls to Microsoft.Toolkit.Services.Markdown. This shouldn't matter too much, since most of this was marked as Internal anyways.

- A lot of Internal Parsing Classes had to be made Public, as they are involved in creating the UI in the MarkdownTextblock, and any other Control based off of the Parser. It appears it was all documented anyway, but it increases the API surface.

- Some RSS Parser and classes moved and changed namespaces, this is mostly fine, but there might need to be some deprecation warnings.

## Other information

I am using this to create an Xamarin Android Markdown TextView, If I create that, I'm more than happy to add it to a future Microsoft.Toolkit.Android.UI.Controls Project :)

This would have to become the Microsoft Community Toolkit though.

WilliamABradley commented:
@nmetulev I was thinking of adding Syntax Highlighting to the Code Blocks, (Using GitHub Style Prefixing a Comment) e.g.

\`\`\`c#
public void Method()
{
}
\`\`\`

makes

```c#
public void Method()
{
}
```

I found a library that could do it, at https://colorcode.codeplex.com/ which I could modify to use Spans for RichTextBlock, and SpannableStrings for Android (Using an interface or abstract class).

I'm not sure how the license is compatible, it is a Microsoft Public License.

I have added an Event to Handle processing Code Block Syntax rendering, would we want to roll built in Syntax Highlighting?

WilliamABradley commented:
I have made some much needed style changes to the default Markdown Style for the Markdown TextBlock.
This style is more aesthetically pleasing, and more closely resembles the styles used in Reddit and Github, in terms of Quote Style, and Code Formatting.

I also rebuilt the Code Inlines, so that they now are more than just a different font family, but actually have a Border, Background, etc.

This should become the default style, but a Compat Style could be provided to anyone who wants to update to this version and keep their old styling, for some reason.

![capture2.png](https://s20.postimg.org/8gu2mkljh/capture2.png)

![Capture.png](https://s20.postimg.org/jf5c57kjx/Capture.png)

michael-hawker commented:
@WilliamABradley wow, cool!  I'll start reviewing this.

I was just needing to do some Markdown parsing on a personal project and thought it'd be great if we could break this parser out.  I wanted to look at some extensions for RichEditBox accepting/returning some basic Markdown, so I think this is a step towards that as well.

WilliamABradley commented:
@michael-hawker @nmetulev I have added built in Syntax Highlighting!

![Syntax Highlighting](https://s20.postimg.org/43j68w571/Capture.png)

Isn't it glorious! Unfortunately there are some missing formats, and things I would like to tweak from ColorCode-Universal, such as the fact that XAML isn't recognised (but xml is), C# Method and Property coloring, etc.

Since I ported the Library in my personal GitHub, and upgraded it there, there are no license conflicts with the UWP Community Toolkit.

Also, @michael-hawker, if you take a look at https://github.com/WilliamABradley/ColorCode-Universal, you could use ColorCode.UWP to make a RichEditBox with Syntax Highlighting, such as a possible `RichEditBoxFormatter`, if that is something you want on your personal project. (We could make a MarkdownEditBox, and leverage the Markdown formatting in ColorCode-Universal to assist writing markdown with Syntax Highlighting, much like Visual Studio and VS Code does).

WilliamABradley commented:
Also Text Selection is a real mess right now and pretty broken (Before this PR too), I think it could be possible to make the Text Selection work across the different internal controls. Perhaps it could be done in a separate PR, but it just isn't there right now. Should we leave text selection off by default?

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MarkdownTextBlock/InitialContent.md:
```diff
@@ -229,7 +229,9 @@ Sometimes you need to preserve indentation, too.  In those cases, you can create
 
 Starting with UWP Community Toolkit v1.4, you can also use GitHub code notification by creating a block surrounded by 3x\` (3 backticks). So:
```
Need to rewrite this sentence so the **So** at the end makes sense

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/MarkdownRenderedEventArgs.cs:
```diff
@@ -18,6 +18,7 @@ namespace Microsoft.Toolkit.Uwp.UI.Controls
     /// rendered.
     /// </summary>
+    [Obsolete("Use Microsoft.Toolkit.Services.Markdown.MarkdownRenderedEventArgs instead.")]
     public class MarkdownRenderedEventArgs : EventArgs
```
This class should just extend the new class. Wouldn't the new class require developers to include the new package as well in order to use the event?

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/MarkdownTextBlock.xaml:
```diff
@@ -3,28 +3,37 @@
+            <SolidColorBrush x:Key="MarkdownInlineCodeBackgroundBrush" Color="#FFDDDFE1" />
+            <SolidColorBrush x:Key="MarkdownQuoteForeground" Color="#FF6E747C" />
         </ResourceDictionary>
         <ResourceDictionary x:Key="Dark">
```
The Dark theme seems broken?

WilliamABradley replied:
You have to set the Foreground color in the Properties Panel, when switching to Dark mode, as it allows you to change it.

WilliamABradley commented:
@nmetulev I was thinking that, however, there is an RSS Parser in the same folder, although that doesn't interact with UI, so I can go that if you want?

nmetulev commented:
Yeah, the RSS parser is internal because it's needed by the bing service. We should probably move it to the same project as the markdown parser and make it public.

WilliamABradley commented:
@nmetulev Awesome, I can get on with that, making Microsoft.Toolkit.Parsers?

Also, should I replace the Sample App's Async Code Coloring Code Panel, with a call to ColorCode-Universal RichTextBlockFormatter, which isn't asynchronous?

nmetulev commented:
Microsoft.Toolkit.Parsers sounds good

What's the benefit of replacing the call with one that's not asynchronous?

WilliamABradley commented:
ColorCode-Universal uses regex and produces HTML, or a Colorized RichTextBlock (So its native), and it has greater support for a larger set of languages.

The current implementation uses a WebView, so it takes longer to load, and doesn't support theme styling, etc. 
It also doesn't seem to support as much of the code.

nmetulev commented:
Ah, I see what you mean. Sounds good, go for it.

michael-hawker commented:
@WilliamABradley I've been bogged down a bit, but hoping I can try this out on something over the holidays.  Will let you know how it goes.

WilliamABradley commented:
@nmetulev I have moved the Markdown and RSS Parsers to `Microsoft.Toolkit.Parsers`, I have also removed the WebView and the JavaScript files for the CodeRenderer, so that is uses ColorCode-Universal.

I removed some regex from `ProcessDocs` in `Sample.cs` as it was hampering some of the `MarkdownTextblock`'s new abilities.

I put back the `MarkdownRenderedEventArgs` so that users don't have to have a dependency on `Microsoft.Toolkit.Parsers`.

WilliamABradley commented:
@hawkerm I ended up adding `ThemeListener` to the UWP Community Toolkit for listening to Theme Changes for the MarkdownTextBlock. I can't seem to find the Issue raised that recommended adding this to the Toolkit, but it is added now.

WilliamABradley commented:
ping @nmetulev I think this is just about done.

nmetulev commented on Microsoft.Toolkit.Parsers/Markdown/Helpers/Common.cs:
```diff
@@ -12,11 +12,14 @@
+    /// <summary>
+    /// Helpers for Markdown.
+    /// </summary>
+    public class Common
```
any reason why this is public?

WilliamABradley replied:
It used to be an issue with creating a Parser for different Libraries, but it might not now.

WilliamABradley replied:
Fixed.

nmetulev commented on Microsoft.Toolkit.Parsers/Markdown/Helpers/DebuggingReporter.cs:
```diff
@@ -12,9 +12,9 @@
 using System.Diagnostics;
 
-namespace Microsoft.Toolkit.Uwp.UI.Controls.Markdown.Helpers
+namespace Microsoft.Toolkit.Parsers.Markdown.Helpers
```
same here re: breaking changes

WilliamABradley replied:
This isn't a breaking change as all of it's containing members were internal.

nmetulev commented on Microsoft.Toolkit.Parsers/Markdown/MarkdownDocument.cs:
```diff
@@ -13,14 +13,15 @@
+using Microsoft.Toolkit.Parsers.Markdown.Parse;
 
-namespace Microsoft.Toolkit.Uwp.UI.Controls.Markdown.Parse
+namespace Microsoft.Toolkit.Parsers.Markdown
```
same here: re breaking changes

WilliamABradley replied:
This isn't a breaking change as all of it's containing members were internal.

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/Microsoft.Toolkit.Uwp.UI.Controls.csproj:
```diff
@@ -11,12 +11,15 @@
   </ItemGroup>
  
   <ItemGroup>
+    <None Include="MarkdownTextBlock\Display\UWPMarkdownRenderer.Blocks.cs" />
```
Don't believe these two are necessary

nmetulev commented on docs/animations/AnimationSet.md:
```diff
@@ -63,50 +61,55 @@ You can change the way how the animation interpolates between keyframes by defin
 | Quintic    | Create an animation that accelerates or decelerates using the formula f(t) = t5                                                                         | ![QuinticEase](https://docs.microsoft.com/en-us/dotnet/framework/wpf/graphics-multimedia/media/quinticease-graph.png)     |
 | Sine       | Creates an animation that accelerates or decelerates using a sine formula                                                                               | ![SineEase](https://docs.microsoft.com/en-us/dotnet/framework/wpf/graphics-multimedia/media/sineease-graph.png)           |
 
-***Note:** EasingType is used only when AnimationSet.UseComposition == false*  
```
Many of these docs changes seem unrelated to this PR. Can you separate these into their own PR?

nmetulev replied:
@Vijay-Nirmal is fixing a lot of these with #1748, and I'd love for him to review this to make sure it wouldn't conflict with his changes.

Vijay-Nirmal replied:
I have changed a lot of things in #1748 to match the [.template.md](https://github.com/Microsoft/UWPCommunityToolkit/raw/master/docs/.template.md) and docs.microsoft.com format. For example, I changed `**Note:**` to `[!NOTE]`. @WilliamABradley What do you think?

> These changes I have made fixed the issues with the documentation pages that caused it not display properly in the MarkdownTextBlock, even with these changes

@WilliamABradley Actually, we can't fix the display issue because [docs.microsoft.com](https://docs.microsoft.com/en-us/contribute/ops-crr/openpublishing/docs/partnerdocs/gfm?toc=/contribute/TOC.json&bc=/contribute/breadcrumb/toc.json) uses [DocFX Flavored Markdown](http://dotnet.github.io/docfx/spec/docfx_flavored_markdown.html). As @nmetulev mention [here](https://github.com/Microsoft/UWPCommunityToolkit/issues/1740#issuecomment-357139872), we have to make a lot of changes to MarkdownTextBlock parser and renderer to fix this issue.

WilliamABradley replied:
@Vijay-Nirmal I have had a look at your changes, it doesn't seem like it would be too hard to merge, Things for you to keep note of, is that currently the MarkdownTextBlock does not support Code Blocks inside of a List Block, in areas of the documentation I removed the Listing areas that used codeblocks, and used bold instead, and other things that didn't present nicely in the MarkdownTextBlock.

There were also several areas of the Documentation where the formatting were broken, like a space at the end ticks of a code block, meaning the rest of the page became the Code Block, as well as other formatting issues.

You are more than welcome to take a copy of the latest commit locally and investigate the changes if you wish (Note, currently the Sample app takes the GitHub master branch copies over the local copies, so you will have to use VS Code MD Preview or some other previewer).

Vijay-Nirmal replied:
I don't think using bold in place of a list is a good idea.

WilliamABradley replied:
@Vijay-Nirmal This is the output:

![Output](https://i.imgur.com/FbfagcG.png)

It looks a lot better than before, when none of the Code Blocks would display properly.

WilliamABradley replied:
This is what it looks like in a list:
![Example](https://i.imgur.com/BuOydh5.png)

Vijay-Nirmal replied:
This is the old look. Actually, I prefer the old format with the new look. Also, How the code block in the list is working properly when the new one doesn't?
![applicationframehost_2018-01-20_11-32-05](https://user-images.githubusercontent.com/24755596/35180401-bc261460-fdd5-11e7-9c69-92d24a99d23c.png)

WilliamABradley replied:
Hmm, that is strange indeed, I will take a look. 

I personally think that the list text should be more intense than the Sample Code/Sample Output text, that way it is more obvious that it is part of that list element.

Maybe bold list items, italic text?

WilliamABradley replied:
Also if you look at it, the formatting is wrong for the code blocks, also the Markdown didn't have any spaces, which is invalid Markdown anyway.

Vijay-Nirmal replied:
We can change Sample Code/Sample Output text to italic or completely remove it.

> Also if you look at it, the formatting is wrong for the code blocks, also the Markdown didn't have any spaces, which is invalid Markdown anyway.

Yes, but it that an unfixable bug?

The reason we cant make the list as bold is that the line describes the example. Using bold in the description is not the good idea. A description can be even more than five lines.

WilliamABradley replied:
After long investigation, the reason the old look works, is because of ProcessDocs() in [Sample.cs](https://github.com/Microsoft/UWPCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp.SampleApp/Models/Sample.cs), which converted all Code Blocks into Quote Blocks.

I will investigate if there is a way I can replicate this structure easily for Code Blocks.

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Display/UWPMarkdownRenderer.cs:
```diff
@@ -0,0 +1,206 @@
+    /// <summary>
+    /// Generates Framework Elements for the UWP Markdown Textblock.
+    /// </summary>
+    public partial class UWPMarkdownRenderer : MarkdownRendererBase
```
It was a more fitting name then XamlRenderer, I can drop the UWP part, it just made it easier to differentiate.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Helpers/Common.cs:
```diff
@@ -12,11 +12,14 @@
+using Microsoft.Toolkit.Parsers.Markdown.Parse;
 
-namespace Microsoft.Toolkit.Uwp.UI.Controls.Markdown.Helpers
+namespace Microsoft.Toolkit.Parsers.Markdown.Helpers
```
This isn't a breaking change as all of it's containing members were internal.

WilliamABradley commented on Microsoft.Toolkit.Services/Core/ExtensionMethods.cs:
```diff
@@ -23,41 +22,6 @@ namespace Microsoft.Toolkit.Services.Core
-        /// <summary>
-        /// Regular expression of HTML tags to remove.
-        /// </summary>
-        private static readonly Regex RemoveHtmlTagsRegex = new Regex(@"(?></?\w+)(?>(?:[^>'""]+|'[^']*'|""[^""]*"")*)>");
```
I will deprecate these.

WilliamABradley commented:
@nmetulev I can fix a lot of the things you request, see the comments I have left.

With the API surface, a lot of it has to expand, as the Parser classes had to become public, so that you can create a Markdown Renderer in another library, I will double check to ensure that only required classes were made public.

I will add all the deprecated classes that were public where needed. 

As for the issues with documentation layout, they have been updated such as `docs/animations/Light.md`, they won't display the changes until after the merge, as the app pulls the docs directly from GitHub.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Core/InternalExtensionMethods.cs:
```diff
@@ -12,7 +12,7 @@
 using System.Text.RegularExpressions;
 
-namespace Microsoft.Toolkit.Services.Core
+namespace Microsoft.Toolkit.Parsers.Core
```
This is an Internal Extension class, so it isn't a breaking change.

WilliamABradley commented:
@QuinnDamerell Would you want some of these changes in Baconit? Syntax Highlighting would be good, but you would probably want to keep your old Xaml Styling. Do you want a compatibility theme?

nmetulev commented on Microsoft.Toolkit.Uwp.Services/Core/DataProviderBase{TConfig,TSchema}.cs:
```diff
@@ -23,7 +21,7 @@ namespace Microsoft.Toolkit.Uwp.Services
     /// <typeparam name="TSchema">Strong typed object to parse the response items into.</typeparam>
     [Obsolete("This class is being deprecated. Please use the .NET Standard Library counterpart found in Microsoft.Toolkit.Services.")]
```
this warning to be changed to the parsers namespace

WilliamABradley replied:
This class is still in the Services namespace.

nmetulev commented on Microsoft.Toolkit.Uwp.Services/Core/ExtensionMethods.cs:
```diff
@@ -31,7 +28,7 @@ public static class ExtensionMethods
         /// <returns>Returns string value.</returns>
         public static string ToSafeString(this object value)
```
The warning should also be added on these methods since they are static extension methods

nmetulev commented on Microsoft.Toolkit.Uwp.UI/DebuggingReporter.cs:
```diff
@@ -10,15 +10,15 @@
     /// </summary>
-    internal interface ILinkRegister
+    [Obsolete("This class is being deprecated. Please use the counterpart in Microsoft.Toolkit.")]
+    public class DebuggingReporter : Toolkit.DebuggingReporter
```
Why is this class public and then deprecated?

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/Microsoft.Toolkit.Uwp.UI.Controls.csproj:
```diff
@@ -11,12 +11,15 @@
   </ItemGroup>
  
   <ItemGroup>
+    <None Include="MarkdownTextBlock\Render\MarkdownRenderer.Blocks.cs" />
```
Are these two includes necessary?

nmetulev commented on Microsoft.Toolkit.Uwp.UI/Helpers/ThemeListener.cs:
```diff
@@ -0,0 +1,136 @@
+using Windows.UI.ViewManagement;
+using Windows.UI.Xaml;
+
+namespace Microsoft.Toolkit.Uwp.UI.Helpers
```
Should this new class have documentation in docs?

WilliamABradley replied:
I could add some documentation for that.

WilliamABradley replied:
Added.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Core/ParseHelpers.cs:
```diff
@@ -10,42 +10,37 @@
+    /// This class offers helpers for Parsing.
     /// </summary>
-    internal class RssParser : IParser<RssSchema>
+    public static class ParseHelpers
```
This checks for tabs as whitespace, which I am not sure is done by the string.IsNullOrWhitespace method.

nmetulev replied:
string.IsNullOrWhiteSpace also checks for tab

WilliamABradley replied:
Hmm, weird, I looked as mscorlib with dotPeek, and it looks like it does check for tab chars, I wonder why this was duplicated? @paulbartrum be advised if you implemented this, will this break anything?

paulbartrum replied:
My two cents: these methods are implementation details of the markdown parser, they should be internal if possible (even if the parser itself is public).

WilliamABradley replied:
Microsoft.Toolkit.Uwp.UI.Controls.Markdown.Render.MarkdownRenderer has a dependency on IsBlankOrWhiteSpace, which is the reason it is public.

WilliamABradley replied:
I think the change in 6e5751be586764db68ecee0a3f75bac91f87aea7 makes it viable to keep it public.

nmetulev replied:
Any reason why the renderer would need access to the parser helpers?

WilliamABradley replied:
It is used in a workaround for UWP, other use cases might need this function.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Rss/RssHelper.cs:
```diff
@@ -18,12 +18,12 @@
     /// Class with utilities for Rss related works.
     /// </summary>
-    internal static class RssHelper
+    public static class RssHelper
```
I thought we were making the RSS parser public, so that others can use them?

nmetulev replied:
Is RssParser not enough to be public in order to be used by others?

WilliamABradley replied:
I'm not sure

WilliamABradley commented on Microsoft.Toolkit.Parsers/Rss/AtomParser.cs:
```diff
@@ -14,14 +14,14 @@
     /// Parser for Atom endpoints.
     /// </summary>
-    internal class AtomParser : BaseRssParser
+    public class AtomParser : BaseRssParser
```
I thought we were making the RSS parser public, so that others can use them?

WilliamABradley commented on Microsoft.Toolkit.Parsers/Rss/BaseRssParser.cs:
```diff
@@ -12,14 +12,14 @@
     /// Base class for Rss Parser(s).
     /// </summary>
-    internal abstract class BaseRssParser
+    public abstract class BaseRssParser
```
I thought we were making the RSS parser public, so that others can use them?

WilliamABradley commented:
@nmetulev I have made most of the changes here, however, I might need help writing up docs for our new Parsers, I am writing up the Markdown Parser docs, and I will do ThemeListener Helpers docs as well.

Would we need a new section in the app for parsers? It will probably be easier to throw them in with the Helpers or Services sections.

nmetulev commented:
For the docs, you can start with the new template that was just added [here](https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md). Let me know if you  have any questions

The parsers should fit in in the Helpers section, don't need to create a new section.

You can't use the SelectionChanged event for what you need?

WilliamABradley commented:
@nmetulev How do you access the SelectionChanged event for a Property?

WilliamABradley commented:
@nmetulev showed me how to do this with the following:
```C#
var sample = DataContext as Sample;
var propDict = sample.PropertyDescriptor.Expando as IDictionary<string, object>;
(propDict["Theme"] as ValueHolder).PropertyChanged += MarkdownTextBlockPage_PropertyChanged;
```

Unfortunately, it wasn't required for my use case in the end, but I put it here for future reference.

WilliamABradley commented:
@nmetulev I believe this is finished, again...

One issue I have discovered is that the MarkdownTextBlock has a slight delay to change Theme, because it fires several property changed events at the same time, causing it to RenderMarkdown about 7-8 times. What would be the best way around this you think?

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        protected override void RenderBlocks(IEnumerable<MarkdownBlock> blockElements, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
Null reference + context_ name should be changed

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        protected override void RenderBlocks(IEnumerable<MarkdownBlock> blockElements, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
+            var blockUIElementCollection_ = context_.BlockUIElementCollection;
```
blockUIElementCollection_ name should be changed.

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+                var frameworkElement = blockUIElementCollection_[i] as FrameworkElement;
+                if (frameworkElement != null)
+                {
+                    if (i == 0)
```
a nice to have comment :) you can do the first item first then do the loop starting from 1 to prevent checking in every loop which will save some performance (not much in this scenario) and if you will do it this way don't forget to check for array length before getting the first item in the array :)

WilliamABradley replied:
This wasn't my code, and I don't want to risk breaking it :)

IbraheemOsama replied:
never mind you can just discard the comment.

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        protected override void RenderListElement(ListBlock element, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
Same thing for null and context_ name

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        protected override void RenderListElement(ListBlock element, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
+            var blockUIElementCollection_ = context_.BlockUIElementCollection;
```
Same here for naming

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        protected override void RenderHorizontalRule(IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
same comment

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        protected override void RenderQuote(QuoteBlock element, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
Same here

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        protected override void RenderCode(CodeBlock element, IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
please apply for all context_ :)

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Inlines.cs:
```diff
@@ -0,0 +1,417 @@
+        /// <param name="context"> Persistent state. </param>
+        protected override void RenderEmoji(EmojiInline element, IRenderContext context)
+        {
+            var context_ = context as InlineRenderContext;
```
Same comment :)

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Inlines.cs:
```diff
@@ -0,0 +1,417 @@
+        /// <param name="context"> Persistent state. </param>
+        protected override void RenderBoldRun(BoldTextInline element, IRenderContext context)
+        {
+            var context_ = context as InlineRenderContext;
```
please apply the same comment on the whole file the name and the null check

IbraheemOsama commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.cs:
```diff
@@ -0,0 +1,206 @@
+        /// <returns>The rich text block</returns>
+        protected RichTextBlock CreateOrReuseRichTextBlock(IRenderContext context)
+        {
+            var context_ = context as UIElementCollectionRenderContext;
```
Null reference

IbraheemOsama commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/CodeRenderer.cs:
```diff
@@ -0,0 +1,164 @@
+// ******************************************************************
+
+using System;
+using System.Text;
```
There are few unused references here :)

WilliamABradley commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/SampleAppMarkdownRenderer.cs:
```diff
@@ -0,0 +1,95 @@
+            var collection = context_.BlockUIElementCollection;
+
+            var viewer = collection.Last() as ScrollViewer;
+            collection.Remove(viewer);
```
This Viewer control is used later, so I need the reference to it anyway.

IbraheemOsama replied:
yeah, I meant only 'collection.Remove(viewer);' because Remove is using RemoveAt underneath

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/MarkdownTextBlock.Methods.cs:
```diff
@@ -0,0 +1,248 @@
+
+            await eventArgs.WaitForDeferrals();
+
+            try
```
No idea, that was from pre-existing code, I only separated MarkdownTextBlock code into partial classes.

WilliamABradley replied:
I'm guessing it is to protect the MarkdownTextBlock from a User Code exception when they override the Image incorrectly.

IbraheemOsama replied:
Ok :)

WilliamABradley commented:
@nmetulev @IbraheemOsama finished the requested changes.

WilliamABradley commented:
@nmetulev @IbraheemOsama @Vijay-Nirmal I have completed the requested changes, but I also threw in DocFX Note formatting for the Sample App, so now it doesn't just replace the text, it presents it as it would in Microsoft Docs!

![DocFX Notes](https://i.imgur.com/qUis5aO.png)

WilliamABradley commented:
Turns out it is currently failing some Markdown Parsing related Unit Tests, I am currently investigating.

WilliamABradley commented:
All issues resolved.

Vijay-Nirmal commented:
Can you change color of Note to match [docs.microsoft.com format](https://docs.microsoft.com/en-us/contribute/ops-crr/openpublishing/docs/partnerdocs/gfm?toc=/contribute/TOC.json&bc=/contribute/breadcrumb/toc.json#markdown-tutorials)?
![chrome_2018-01-20_15-12-48](https://user-images.githubusercontent.com/24755596/35182124-bafa0db6-fdf4-11e7-8db2-e2c30f77bedf.png)
And in dark theme
![chrome_2018-01-20_15-12-27](https://user-images.githubusercontent.com/24755596/35182130-c88c8aee-fdf4-11e7-8624-036007df429a.png)

WilliamABradley commented:
I followed DocFX’s Documentation, but that looks better, and it has more diverse colours.

Thanks,
William

________________________________
From: Vijay Nirmal <notifications@github.com>
Sent: Saturday, January 20, 2018 10:46:19 PM
To: Microsoft/UWPCommunityToolkit
Cc: William Bradley; Mention
Subject: Re: [Microsoft/UWPCommunityToolkit] Markdown + RSS + JSON Parser for .NET Standard, MarkdownTextBlock Style Improvements, Syntax Highlighting (#1650)


Can you change color of Note to match docs.microsoft.com format<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fcontribute%2Fops-crr%2Fopenpublishing%2Fdocs%2Fpartnerdocs%2Fgfm%3Ftoc%3D%2Fcontribute%2FTOC.json%26bc%3D%2Fcontribute%2Fbreadcrumb%2Ftoc.json%23markdown-tutorials&data=02%7C01%7C%7C0ecd82d4407c46475c0608d55feaa944%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636520383820891582&sdata=MyJ8s2%2F%2BLzGbF4WPw3DOIGE5J7UBHfwpXfTrJBugkxg%3D&reserved=0>?
[chrome_2018-01-20_15-12-48]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F24755596%2F35182124-bafa0db6-fdf4-11e7-8db2-e2c30f77bedf.png&data=02%7C01%7C%7C0ecd82d4407c46475c0608d55feaa944%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636520383820891582&sdata=8VpJxlF%2BGZfRzLZee4tAN4HFDapkjfHENBcGaMHhLb8%3D&reserved=0>
And in dark theme
[chrome_2018-01-20_15-12-27]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F24755596%2F35182130-c88c8aee-fdf4-11e7-8624-036007df429a.png&data=02%7C01%7C%7C0ecd82d4407c46475c0608d55feaa944%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636520383820891582&sdata=wGxaVMPjD0yvmiTp5wEn%2FAayOV%2BH472CkYR5ilkFwbw%3D&reserved=0>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2FUWPCommunityToolkit%2Fpull%2F1650%23issuecomment-359159475&data=02%7C01%7C%7C0ecd82d4407c46475c0608d55feaa944%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636520383820891582&sdata=dYsrQl50ZAe84m8BUSCTBiJ0onKB%2FqFG%2BXvL%2FAmH2go%3D&reserved=0>, or mute the thread<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFv1KToqG-8f6eSO-QRoAruekYX_m7jFks5tMbXrgaJpZM4Qqqol&data=02%7C01%7C%7C0ecd82d4407c46475c0608d55feaa944%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636520383820891582&sdata=UYU%2BcRn4FDrj%2BSMi0%2FjVLckC%2BT2xaTbRcSUQ0KSD75M%3D&reserved=0>.

WilliamABradley commented:
@Vijay-Nirmal CodeBlocks in ListBlocks are now only semi-broken instead of fully broken :) Currently Tab Chars are being swallowed by the List Logic, so I have to figure out how to rescue them.

![CodeBlock in ListBlock](https://i.imgur.com/lJbaQX9.png)

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Blocks/ListBlock.cs:
```diff
@@ -169,8 +169,8 @@ internal static ListBlock Parse(string markdown, int start, int maxEnd, int quot
                         else
                         {
-                            // Inline text.
-                            AppendTextToListItem(currentListItem, markdown, lineInfo.FirstNonWhitespaceChar, lineInfo.EndOfLine, true);
```
@paulbartrum This is a bug I have found where a ListBlockItem will swallow any additional formatting Whitespace, such as used in CodeBlocks.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Blocks/ListBlock.cs:
```diff
@@ -169,8 +169,8 @@ internal static ListBlock Parse(string markdown, int start, int maxEnd, int quot
-                            // Inline text.
-                            AppendTextToListItem(currentListItem, markdown, lineInfo.FirstNonWhitespaceChar, lineInfo.EndOfLine, true);
+                            // Inline text. Ignores the 4 spaces that are used to continue the list.
+                            AppendTextToListItem(currentListItem, markdown, lineInfo.StartOfLine + 4, lineInfo.EndOfLine, true);
```
@paulbartrum This is a bug fix for where a ListBlockItem will swallow any additional formatting Whitespace, such as used in CodeBlocks.

WilliamABradley replied:
Turns out it was a little bit more complicated then that, added more workaround in the next commit.

WilliamABradley commented:
@Vijay-Nirmal @nmetulev I have adjusted the DocFX Notation formatting to be in line with docs.microsoft.com.

![Notation](https://i.imgur.com/1Np5yVh.png)

I have also restored the List style examples in Animations, after I fixed the bug that was causing formatting issues.

![Animations Examples](http://i.imgur.com/LCOECcv.jpg)

There is an issue where Several Multi-Line Quotes separated by NewLines merge into one large Quote, but that is a PR for another day (It was like that before this PR) :)

I think that's a wrap!

nmetulev commented on docs/parsers/RSSParser.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: RSS Parser
+author: williamabradley
+ms.date: 01/19/2018
```
remove date

nmetulev commented on docs/parsers/RSSParser.md:
```diff
@@ -0,0 +1,30 @@
+
+You can see this in action in [UWP Community Toolkit Sample App](https://www.microsoft.com/store/apps/9NBLGGH4TLCQ).
+
+## Requirements
```
Requirements need to show the namespace and nuget package (see template: https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md)

WilliamABradley replied:
Done.

nmetulev commented on docs/parsers/RSSParser.md:
```diff
@@ -0,0 +1,30 @@
+
+The [RssParser](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.parsers.rss.rssparser) class allows you to parse a RSS content String into a RSS Schema.
+
+| Class | Purpose |
```
Properties should be in a section called Properties (see template: https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md)

nmetulev commented on docs/parsers/RSSParser.md:
```diff
@@ -0,0 +1,30 @@
+# RSS Parser
+
+The [RssParser](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.parsers.rss.rssparser) class allows you to parse a RSS content String into a RSS Schema.
+
```
Is there any code snipet you can add here?

WilliamABradley replied:
Done.

nmetulev commented on docs/parsers/MarkdownParser.md:
```diff
@@ -0,0 +1,48 @@
+---
+title: Markdown Parser
+author: williamabradley
+ms.date: 01/19/2018
```
remove date

nmetulev commented on docs/parsers/MarkdownParser.md:
```diff
@@ -0,0 +1,48 @@
+
+You can see this in action in [UWP Community Toolkit Sample App](https://www.microsoft.com/store/apps/9NBLGGH4TLCQ).
+
+## Requirements
```
Please follow requirements section structure from template https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md

WilliamABradley replied:
Done.

nmetulev commented on docs/parsers/MarkdownParser.md:
````diff
@@ -0,0 +1,48 @@
+```csharp
+string md = "This is **Markdown**";
+MarkdownDocument Document = new MarkdownDocument();
+Document.Parse(md);
````
Maybe add few lines here to demonstrate the output and how it can be used?

WilliamABradley replied:
Done.

nmetulev commented on docs/parsers/JsonParser.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: JSON Parser
+author: williamabradley
+ms.date: 01/19/2018
```
Remove date

nmetulev commented on docs/parsers/JsonParser.md:
```diff
@@ -0,0 +1,30 @@
+
+The [JsonParser](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.parsers.json.jsonparser) class allows you to parse a JSON content String into a JSON Schema.
+
+| Class | Purpose |
```
Add properties heading per template: https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md

nmetulev commented on docs/parsers/JsonParser.md:
```diff
@@ -0,0 +1,30 @@
+# JSON Parser
+
+The [JsonParser](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.parsers.json.jsonparser) class allows you to parse a JSON content String into a JSON Schema.
+
```
Is there a code snippet you can add?

nmetulev commented on docs/parsers/JsonParser.md:
```diff
@@ -0,0 +1,30 @@
+
+You can see this in action in [UWP Community Toolkit Sample App](https://www.microsoft.com/store/apps/9NBLGGH4TLCQ).
+
+## Requirements
```
Requirements needs to follow template structure: https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md

WilliamABradley replied:
Done.

nmetulev commented on docs/helpers/ThemeListener.md:
```diff
@@ -0,0 +1,56 @@
+---
+title: Theme Listener
+author: williamabradley
+ms.date: 01/19/2018
```
Remove date

nmetulev commented on docs/helpers/ThemeListener.md:
```diff
@@ -0,0 +1,56 @@
+
+## Requirements (Windows 10 Device Family)
+
+| [Device family](http://go.microsoft.com/fwlink/p/?LinkID=526370) | Universal, 10.0.14393.0 or higher |
```
Need to add a link to the nuget package

WilliamABradley replied:
Done.

nmetulev commented on docs/controls/SlidableListItem.md:
````diff
@@ -86,10 +81,9 @@ If you use **SlidableListItem** in a **ListView** with the **ItemClick** event,
 ```
 
-## <a name="swipe"></a> Moving to SwipeControl
+## Moving to SwipeControl
````
There are external links that would be broken by changing this, please revert back

nmetulev commented on docs/controls/MarkdownTextBlock.md:
```diff
@@ -1,7 +1,7 @@
-author: nmetulev
-ms.date: 08/20/2017
+author: quinndamerell, deltakosh, tipa, haefele, avknaidu, nmetulev, shenchauhan, vijay-nirmal, pedrolamas, williamabradley
+ms.date: 12/19/2017
```
please remove date

nmetulev commented on docs/controls/HamburgerMenu.md:
````diff
@@ -143,10 +143,9 @@ namespace HamburgerSample
-
 ```
 
-## <a name="navview"></a> Moving to NavigationView
````
there are external links that would be broken with this change, please revert back

nmetulev commented on docs/animations/ParallaxService.md:
```diff
@@ -43,14 +43,14 @@ MyUIElement.SetValue(ParallaxService.HorizontalMultiplierProperty, 0.5);
 
 [ParallaxService Sample Page Source](https://github.com/Microsoft/UWPCommunityToolkit/tree/master/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ParallaxService). You can see this in action in [UWP Community Toolkit Sample App](https://www.microsoft.com/store/apps/9NBLGGH4TLCQ)
 
-## <a name="parallaxview"></a> Moving to ParallaxView
```
There are external links that would be broken with this change, please revert back

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/JsonParser/JsonParserPage.xaml:
```diff
@@ -0,0 +1,11 @@
+﻿<Page
```
Sample without sample doesn't make sense. Remove from sample app.

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/RssParser/RssParserPage.xaml:
```diff
@@ -0,0 +1,11 @@
+﻿<Page
```
sample without sample doesn't make sense, remove from sample app

WilliamABradley replied:
Added Sample.

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ThemeListener/ThemeListenerPage.xaml:
```diff
@@ -0,0 +1,26 @@
+﻿<Page
```
Remove this sample. This should be added once the sample app supports dark theme

WilliamABradley commented:
@nmetulev are we able to fix the external links? We should **_NOT_** be embedding HTML tags into Markdown, it breaks presentation in most Markdown Renderers, including the MarkdownTextBlock. This is also not the intended use of Markdown markup.

```
## <a name="swipe"></a> Moving to SwipeControl 

...

[Moving to SwipeControl](#swipe)
```

This can be just as easily represented using:

```
## Moving to SwipeControl 

...

[Moving to SwipeControl](#moving-to-swipecontrol)
```

WilliamABradley commented:
Another question, can we show the Documentation in the Sample App without the Sample Page? Removing the type property from **sample.json** means that clicking on the page causes nothing to load.

Vijay-Nirmal commented:
Add `Microsoft.Toolkit.Parsers` in Nuget Packages table in [readme.md](https://github.com/Microsoft/UWPCommunityToolkit/blob/master/readme.md)

nmetulev commented:
On the external links, let's leave the changes you have. The links will still work, just won't scroll down to the right section which is ok.

On documentation in the sample app, we currently don't have a way to view documentation of features that do not have a sample. We would need to build a section in the app to browse documentation (which I think is a good idea). It is out of scope for this PR though, we should create a new issue.

Going back on the parsers, I reviewed them in more details and I don't see a reason why the JSON parser should be public. It does not provide any value for developers over other parsers and I'd prefer it to be internal.

WilliamABradley commented:
@nmetulev I agree with about the JSON Parser, I changed back the external links, and I added support to the Markdown Parser, so that they get parsed into inlines and ignored, so they no longer appear in the MarkdownTextBlock.

In the future, we should try to remove them (Although docs.microsoft.com doesn't support linking to headers like GitHub does).

This is necessary because the MarkdownTextBlock might be parsing other sources of Markdown that might have these links.

*****

I would like to keep the Samples for RSSParser and ThemeListener in the Sample App even though RSSParser doesn't have any sample on it's page, and ThemeListener's Sample doesn't currently work. These still have Documentation Pages that are important. Like the Analytics Page.

WilliamABradley commented:
@nmetulev Changes have been made according to Review.

WilliamABradley commented:
@nmetulev I will ensure other active MarkdownTextBlock PRs properly handle the transition after this PR goes through.

nmetulev commented on Microsoft.Toolkit/DebuggingReporter.cs:
```diff
@@ -12,9 +12,12 @@
+    /// <summary>
+    /// Reports an error during debugging.
+    /// </summary>
+    public class DebuggingReporter
```
This class should not be public. It is only needed for debugging.

WilliamABradley commented:
@nmetulev I have fixed the crash in the MarkdownTextBlock.

ThemeListener is a required component for MarkdownTextBlock Syntax highlighting, since we are adding it, I thought it would be good to at least have documentation for it in the sample app, even if the Sample doesn't work. (I could remove the asterisk line, and technically it would be correct, although you wouldn't see the value change 🙂)

#1639 shouldn't be affected by the changes as Github is pretty good at shifting code in an auto-merge. 

#1634 hasn't been updated since November.

nmetulev commented:
Thanks. Agree that the ThemeListener sample is good, but since you are adding the dark theme in your other PR, I'd rather add the sample there - it will make it easier to validate and test.

WilliamABradley commented:
@nmetulev I disabled the ThemeListener Sample.

Vijay-Nirmal commented:
@WilliamABradley Can you revert the doc changes? because it will cause conflict with #1748 and also, these changes have nothing to do with this PR.

WilliamABradley commented:
@Vijay-Nirmal These changes in Documentation fix a number of issues with the presentation in the Sample App (A big part of this PR) as well as general issues. 

The conflicts don't run too deep, so it shouldn't be too difficult for you to merge. Also your changes only affect Animations, while mine affect all Documentation.

WilliamABradley commented:
There are a lot of issues which I have fixed (Which you can see for yourself by looking at the changes to the Documentation I have made in this PR). 

If you want to see the changes in-app, and what you might need to fix, you can take a copy of this PR's latest commit to test it yourself if you want.

This is 3 months of PR, so that trumps two weeks of changes 😄 

I can revert the changes to Animations, minus the fixes to make it present properly if you want. The rest of the documentation being reverted is another can of worms.

Vijay-Nirmal commented on docs/animations/AnimationSet.md:
````diff
@@ -71,6 +69,7 @@ You can change the way how the animation interpolates between keyframes by defin
 
     **Sample Code**
+
     ```csharp
````
Is this is a fix for MarkdownTextBlock issue or is this a general improvement?

WilliamABradley replied:
Both, that is proper Markdown. Markdown requires a new line to be two returns long.

Vijay-Nirmal replied:
But displaying property without this fix!

WilliamABradley replied:
It isn't proper, and it doesn't display properly in the Sample App.

Vijay-Nirmal replied:
Displayed properly in old sample app

Vijay-Nirmal commented on docs/animations/AnimationSet.md:
```diff
@@ -43,8 +43,6 @@ anim.Start();
 ## Properties
 
-
-
```
General improvement. So you can revert it.

WilliamABradley replied:
Will you apply this fix to all docs it affects?

Vijay-Nirmal replied:
Yes

Vijay-Nirmal commented on docs/animations/Blur.md:
```diff
@@ -39,15 +39,12 @@ await MyUIElement.Blur(value: 5, duration: 2500, delay: 250).StartAsync();  //Bl
 
 ![Blur Behavior animation](../resources/images/Animations/Blur/Sample-Output.gif)
 
-## Properties
```
I added Properties section in my PR. So you can revert it.

Vijay-Nirmal commented on docs/animations/CompositionAnimations.md:
````diff
@@ -23,7 +23,7 @@ These XAML elements enable developer to specify composition animations directly
     <animations:ScalarKeyFrame Key="0.5" Value="0.0"></animations:ScalarKeyFrame>
 </animations:ScalarAnimation>
- ```
+```
````
Fixed in my PR. So you can revert it.

WilliamABradley replied:
Auto-merge will handle this.

Vijay-Nirmal commented on docs/animations/ParallaxService.md:
````diff
@@ -50,7 +50,7 @@ However, the way the two are used is different. Unlike the ParallaxService, the
 Here is an example of using the ParallaxView
 
-```csharp
+```xaml
````
I will fix this in my PR. So you can revert it.

WilliamABradley replied:
There isn't any need.

Vijay-Nirmal replied:
My point is you can revert all the doc changes.

Vijay-Nirmal commented on docs/controls/AdaptiveGridView.md:
```diff
@@ -15,13 +15,11 @@ If there are not enough items to fill one row, the control will stretch the item
-          ItemHeight="200"
-          DesiredWidth="300"
-          ItemTemplate="{StaticResource PhotosTemplate}">
+    ItemHeight="200"
```
Is this is a fix for MarkdownTextBlock issue or is this a general improvement?

Vijay-Nirmal commented on docs/controls/Carousel.md:
```diff
@@ -71,16 +69,14 @@ namespace Windows.UI.Xaml.Controls
-        }
-
+/// <summary>
+/// Gets or sets depth of non Selected Items
```
You can revert this. I have changed the whole section.

Vijay-Nirmal commented on docs/controls/DockPanel.md:
```diff
@@ -37,16 +35,15 @@ You can set DockPanel LastChildFill property to true if you want the last item a
-      xmlns:local="using:Microsoft.Toolkit.Uwp.SampleApp.SamplePages"
-      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
-      mc:Ignorable="d" >
+  xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
```
You can revert this. I have changed the whole section.

Vijay-Nirmal commented on docs/controls/HamburgerMenu.md:
```diff
@@ -9,7 +9,7 @@ keywords: windows 10, uwp, uwp community toolkit, uwp toolkit, HamburgerMenu, xa
 
 > [!NOTE]
-The HamburgerMenu is deprecated and will be removed in a future major release. Please use the [NavigationView](https://docs.microsoft.com/en-us/windows/uwp/controls-and-patterns/navigationview) available in the Fall Creators Update. Read the [Moving to NavigationView](#navview) section for more info.
+The HamburgerMenu is deprecated and will be removed in a future major release. Please use the [NavigationView](https://docs.microsoft.com/en-us/windows/uwp/controls-and-patterns/navigationview) available in the Fall Creators Update. Read the [Moving to NavigationView](#moving-to-navigationview) section for more info.
```
This won't work in docs.microsoft.com. So you must revert it.

Vijay-Nirmal replied:
@WilliamABradley You have made the same changes in other docs also.

Vijay-Nirmal commented on docs/controls/StaggeredPanel.md:
```diff
@@ -8,11 +8,6 @@ keywords: windows 10, uwp, uwp community toolkit, uwp toolkit, StaggeredPanel
-<!-- Use below format to display note
-> [!NOTE]
-Some note
--->
```
I have removed this in my PR. So you can revert it.

Vijay-Nirmal commented on docs/extensions/ListViewBase.md:
```diff
@@ -8,86 +8,84 @@ keywords: windows 10, uwp, uwp community toolkit, uwp toolkit, ListViewBase, ext
 The AlternateColor property provides a way to assign a background color to every other item.
 
-> The <a href="https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.listviewbase#Windows_UI_Xaml_Controls_ListViewBase_ContainerContentChanging" target="_blank">ContainerContentChanging</a> event used for this extension to work, will not be raised when the ItemsPanel is replaced with another type of panel than ItemsStackPanel or ItemsWrapGrid. 
+> [!WARNING] The [ContainerContentChanging](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.listviewbase#Windows_UI_Xaml_Controls_ListViewBase_ContainerContentChanging) event used for this extension to work, will not be raised when the ItemsPanel is replaced with another type of panel than ItemsStackPanel or ItemsWrapGrid. 
```
I have fixed this in my PR. So you can remove it.

Vijay-Nirmal commented on docs/extensions/SurfaceDialTextboxHelper.md:
```diff
@@ -26,13 +27,11 @@ This property enables the MinValue and MaxValue limits of the Textbox. These val
 
-
-**NOTE:** Windows Anniversary Update (10.0.14393.0) is needed to support correctly this helper.
+> [!NOTE] Windows Anniversary Update (10.0.14393.0) is required for this helper to work correctly.
```
Fixed this in my PR. So you can remove it.

WilliamABradley commented:
@Vijay-Nirmal if you have made duplicate changes in your PR, then auto-merge will be able to do most of the work. A lot of these fixes are small, so any conflicts won't be hard to handle.

Vijay-Nirmal commented:
@WilliamABradley My point is you can revert all the doc changes. If anything needs to be fixed I will fix it in my PR. Easy for you as well as for me.

WilliamABradley commented:
@Vijay-Nirmal  What I am saying, is that there are already numerous changes to documentation already committed, and for you to do it again likely will mean things get missed, and it will take extra time, and also require extra commits for me to undo things.

WilliamABradley commented on Microsoft.Toolkit.Uwp.SampleApp/Models/Sample.cs:
````diff
@@ -236,25 +240,8 @@ private string ProcessDocs(string docs)
 
-            // Need to do some cleaning
-            // Rework code tags
-            var regex = new Regex("```(xaml|xml|csharp)(?<code>.+?)```", RegexOptions.Singleline);
````
@Vijay-Nirmal This Regex was the reason some of these documentation issues still worked in the old Sample App, it converted all Code Blocks into a different block type, which prevented Syntax Highlighting or anything else.

It has been removed so we can use it properly now.

nmetulev commented on docs/controls/MarkdownTextBlock.md:
````diff
@@ -119,6 +139,36 @@ private async void MarkdownText_OnImageResolving(object sender, ImageResolvingEv
+block.SetRenderer<InheritedUWPMarkdownRenderer>();
+```
+
+This might require intimate knowledge of the implementation of the `UWPMarkdownRenderer`, take a look at the following:
````
Change name to MarkdownRenderer

WilliamABradley replied:
@nmetulev Done.

michael-hawker commented on Microsoft.Toolkit.Parsers/Core/SchemaBase.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Microsoft.Toolkit.Parsers
+{
+    /// <summary>
+    /// Strong typed schema base class.
```
Could be handy to have more details here on what this is used for.

WilliamABradley replied:
It is used by the RSS Parser, and some of the Services like Bing? It is a base class for Schema based parsers.

michael-hawker commented on Microsoft.Toolkit.Parsers/Core/StringValueAttribute.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Microsoft.Toolkit.Parsers.Core
+{
+    /// <summary>
+    /// StringValue attribute.
```
StringValue attribute used by...?  extra pointers would be helpful here.

WilliamABradley replied:
It is used by the Bing Service, I believe.

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/Enums/InlineParseMethod.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace Microsoft.Toolkit.Parsers.Markdown.Enums
+{
+    internal enum InlineParseMethod
```
How is this different than `MarkdownInlineType`?

WilliamABradley replied:
`InlineParseMethod` is used for identifying Inlines with TripChars, the other is for structure identification.
An inline type might have no `MarkdownInlineType`, private Inlines like BoldItalicInline which converts itself into a BoldInline, it is a workaround.

This stuff has always been here, I might clean it up / remove it in a new PR (Like with #1767 where I will likely remove the enums altogether).

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/Helpers/LineInfo.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Microsoft.Toolkit.Parsers.Markdown.Helpers
+{
+    internal class LineInfo
```
Is this for tracking block positions?

If so, could be useful to expose for locating/highlighting text based on parsing?

WilliamABradley replied:
Again, this should be work for a PR based #1767, where we open up the internals of the Parser for tinkering.
 
Since you couldn't really do much with this info right now, and it would require considerable more work to expose, leave it for now.

This PR is big enough as it is :)

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/Inlines/BoldItalicTextInline.cs:
```diff
@@ -49,14 +51,14 @@ internal static void AddTripChars(List<Common.InlineTripCharHelper> tripCharHelp
             }
 
-            if (markdown == null || markdown.Length < 6)
+            if (markdown == null || markdown.Length < 6 || start + 6 >= maxEnd)
```
What's the '6' for?

WilliamABradley replied:
It ensures that the length of the string can possibly fit two sets of "\*\*\*", which is required for Bold and Italic parsing. Without these, creating a BoldItalicInline will crash if prepended or appended to the text, due to the substring calls.

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/MarkdownDocument.cs:
```diff
@@ -13,14 +13,17 @@
+    /// Initialize an instance and call <see cref="Parse(string)"/> to parse the Raw Markdown Text.
     /// </summary>
-    internal class MarkdownDocument : MarkdownBlock
+    public class MarkdownDocument : MarkdownBlock
```
Could be nice if the constructor can also take a string and just pass that to the Parse method.

Is there a reason this doesn't have the IParser interface too?

WilliamABradley replied:
IParser is used for schema based parsers, like some XML, and RSS. Markdown doesn't really fit this.

The Constructor with no string input, and a void Parse method, mimics the style of XDocument and XmlDocument parsing in .NET itself, so I think it should stay that way.

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/MarkdownDocument.cs:
```diff
@@ -365,4 +368,4 @@ public override string ToString()
             return string.Join("\r\n", Blocks);
```
Might not have gotten to it yet.  Is there a test that checks that the output of the MarkdownDocument toString is the same as what was passed in to Parse?

WilliamABradley replied:
No, that was before my PR :)

michael-hawker replied:
Might be good to think about adding another test then that has a sample markdown text blob, gets read by the parser, and then the document gets serialized back to markdown and compare the results?

WilliamABradley replied:
There are numerous cases where this isn't done, or the returned ToString isn't the same is the input. Examples are the `CodeBlock` where it just returns the text. `HorizontalRuleBlock` only returns one kind of raw line, when there are several ways to do it, etc.

It could be done, but it would take some considerable reworking, possibly for a future PR like #1767 or future PRs.

michael-hawker commented on Microsoft.Toolkit.Parsers/Markdown/MarkdownElement.cs:
```diff
@@ -10,12 +10,12 @@
     /// </summary>
-    internal abstract class MarkdownElement
+    public abstract class MarkdownElement
     {
```
Is there another block type we could use here on all elements?  Trying to think if that'd be nicer than having to check class instances.

Also wondering if we want a common 'Text' property or if toString is sufficient.  Or maybe 'DisplayText' which would have the text without the markup?

WilliamABradley replied:
Currently the Text Property belongs to IInlineLeaf, as not all Inlines have text. And Blocks (Which inherit from MarkdownElement definitely don't have a text property.

michael-hawker commented on Microsoft.Toolkit.Parsers/Rss/AtomParser.cs:
```diff
@@ -14,9 +14,9 @@
-using Microsoft.Toolkit.Services.Core;
+using Microsoft.Toolkit.Parsers.Core;
 
-namespace Microsoft.Toolkit.Services.Rss
```
Was this internal usage before too?

WilliamABradley replied:
Yep

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/SampleAppMarkdownRenderer.cs:
```diff
@@ -0,0 +1,302 @@
+            {
+                case "CSHARP":
+                case "CS":
+                    language = "C#";
```
Should this not be in the parser somewhere vs the renderer?

WilliamABradley replied:
This is just for the header of the Sample App's Code Block that has a header specifying the language of the Code, just like docs.microsoft.com.

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/SampleAppMarkdownRenderer.cs:
```diff
@@ -0,0 +1,302 @@
+            }
+        }
+
+        private const string NoteGlyph = "\uE946";
```
Should this be configurable?  Or is it part of the Common Mark spec?

WilliamABradley replied:
This is only for the Sample App, it is part of DocFX, so it isn't included in our CommonMark-ish parser.

WilliamABradley commented:
ping @nmetulev All dependent PRs are now merged :)

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/SampleAppMarkdownRenderer.cs:
```diff
@@ -0,0 +1,302 @@
+
+namespace Microsoft.Toolkit.Uwp.SampleApp.Controls
+{
+    internal class SampleAppMarkdownRenderer : MarkdownRenderer
```
Since this is specific to the sample app and some stuff we do there.  Maybe preface this class with details on what this is doing and why?

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/Models/Sample.cs:
```diff
@@ -169,53 +175,52 @@ public async Task<(string contents, string path)> GetDocumentationAsync()
-                        }
-                    }
-                }
+                _cachedDocumentation = await StorageFileHelper.ReadTextFromLocalCacheFileAsync(filename);
```
Did we want to keep the !DEBUG so we read the latest from dev?

WilliamABradley replied:
The latest is only read in Release mode, a.k.a. in the Store. If it is debug mode, we want to get the local docs, that way we can see changes when we edit the documentation locally.

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MarkdownTextBlock/InitialContent.md:
```diff
@@ -250,9 +250,9 @@ Sometimes you need to preserve indentation, too.  In those cases, you can create
+Starting with UWP Community Toolkit v1.4, you can also use GitHub code notification by creating a block surrounded by 3x\` (3 backticks). This can also be used with Language Identifiers on the entering backticks, such as:
 
-\`\`\`
+\`\`\`java
```
use csharp instead as we don't code in java for UWP? ;)

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MarkdownTextBlock/MarkdownTextBlockPage.xaml.cs:
```diff
@@ -79,5 +84,15 @@ private void SetInitalText(string text)
         }
+
+        // Custom Code Block Renderer
+        private void MarkdownText_CodeBlockResolving(object sender, CodeBlockResolvingEventArgs e)
```
This is basically a parser event (I can't see where it's raised from yet).  Should it be `ParseCustomCodeBlock`?

I didn't see the term `Resolving` used elsewhere in the parser, so just want to make sure we're consistent and not introducing a bunch of terms.

WilliamABradley replied:
It is used by `ImageResolving` in MarkdownTextBlock.Events.cs.

michael-hawker replied:
Commented on it down there when I found it, this is fine. :)

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MarkdownTextBlock/MarkdownTextBlockPage.xaml.cs:
```diff
@@ -79,5 +84,15 @@ private void SetInitalText(string text)
+            if (e.CodeLanguage == "CUSTOM")
+            {
+                e.Handled = true;
+                e.InlineCollection.Add(new Run { Foreground = new SolidColorBrush(Colors.Red), Text = e.Text, FontWeight = FontWeights.Bold });
```
Should this example do something with the text like look for a keyword and only make that part red?  It might be a more useful example that way.

WilliamABradley replied:
I don't really want to delve into Regex for this one, it is a simple sample after all.

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/samples.json:
```diff
@@ -615,6 +615,13 @@
         "Icon": "/SamplePages/AdvancedCollectionView/AdvancedCollectionView.png",
         "DocumentationUrl": "https://raw.githubusercontent.com/Microsoft/UWPCommunityToolkit/master/docs/helpers/AdvancedCollectionView.md"
       }
+      //{
```
As @nmetulev said, move to other PR.

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/samples.json:
```diff
@@ -697,6 +704,26 @@
+        "Name": "Markdown Parser",
+        "Type": "MarkdownParserPage",
+        "About": "The Markdown Parser allows you to parse a Markdown String into a Markdown Document, and then Render it with a Markdown Renderer.",
+        "Icon": "/Assets/Helpers.png",
```
@nmetulev do we have some designer time before 2.2 to get images for these (and other things like composition brushes)?

WilliamABradley replied:
We could create a cleanup PR, there are several VS complaints from other Toolkit items that need to be fixed too.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/MarkdownTextBlock.Events.cs:
```diff
@@ -0,0 +1,98 @@
+        /// The default implementation is basically <code>new BitmapImage(new Uri(e.Url));</code>.
+        /// <para/>You must set <see cref="ImageResolvingEventArgs.Handled"/> to true in order to process your changes.
+        /// </summary>
+        public event EventHandler<ImageResolvingEventArgs> ImageResolving;
```
I see the Resolving now, this already existed eh?  If so, we can leave the other name as is.

WilliamABradley replied:
Yep.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,432 @@
+                switch (element.Style)
+                {
+                    case ListStyle.Bulleted:
+                        bullet.Text = "•";
```
Should we use escape sequence vs. explicit unicode character?  Should what this character is be configurable?

WilliamABradley replied:
This was an existing implementation.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Blocks.cs:
```diff
@@ -0,0 +1,432 @@
+            {
+                FontFamily = CodeFontFamily ?? FontFamily,
+                Foreground = brush,
+                LineHeight = FontSize * 1.4
```
Magic Number?

michael-hawker replied:
As discussed offline, was here before, we can worry about another day.

WilliamABradley replied:
This was an existing implementation.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Dimensions.cs:
```diff
@@ -0,0 +1,209 @@
+    /// <summary>
+    /// Measurement Properties for elements in the Markdown.
+    /// </summary>
+    public partial class MarkdownRenderer
```
Are these are the same properties on the control?  Should there be a common interface?

WilliamABradley replied:
Not quite, the ones on the control are dependency properties.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Inlines.cs:
```diff
@@ -0,0 +1,493 @@
+                throw new RenderContextIncorrectException();
+            }
+
+            // HACK: Superscript is not allowed within a hyperlink.  But if we switch it around, so
```
@nmetulev is this a platform issue?  Are we tracking it somewhere?

WilliamABradley replied:
This is a long standing issue I believe.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/MarkdownTextBlock/Render/MarkdownRenderer.Inlines.cs:
```diff
@@ -0,0 +1,493 @@
+
+            var border = new Border
+            {
+                Padding = new Thickness(0, 0, 0, paragraph.FontSize * 0.2),
```
Magic Number?

WilliamABradley replied:
I didn't change anything here, it was moved.

michael-hawker replied:
As discussed offline, was here before, we can worry about another day.

WilliamABradley replied:
This was an existing implementation.

michael-hawker commented on UWP Community Toolkit.sln:
```diff
@@ -50,12 +50,12 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "UnitTests.Notifications.Win
 	EndProjectSection
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Toolkit.Uwp.Notifications.JavaScript", "Microsoft.Toolkit.Uwp.Notifications.JavaScript\Microsoft.Toolkit.Uwp.Notifications.JavaScript.csproj", "{94F8D702-3A9D-4CFB-85C9-79FC5DBD8B22}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Toolkit.Uwp.Notifications.JavaScript", "Microsoft.Toolkit.Uwp.Notifications.JavaScript\Microsoft.Toolkit.Uwp.Notifications.JavaScript.csproj", "{94F8D702-3A9D-4CFB-85C9-79FC5DBD8B22}"
```
Do we know why this guid changed?

WilliamABradley replied:
No idea? Perhaps pulling master?

michael-hawker commented on docs/controls/WrapPanel.md:
```diff
@@ -44,15 +42,14 @@ When the Orientation is Vertical, HorizontalSpacing adds uniform spacing between
-      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
-      xmlns:wrapPanel="using:Microsoft.Toolkit.Uwp.UI.Controls.WrapPanel"
-      mc:Ignorable="d">
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
```
Lots of whitespace changes, these don't lineup anymore. :(

WilliamABradley replied:
It was Visual Studio, It is probably better because that means these will take less horizontal space.

michael-hawker replied:
I just like my xaml attributes to line up, but maybe I'm weird that way...

WilliamABradley replied:
It would a major pain to fix all of the code comments :(

michael-hawker commented on UnitTests/Markdown/Parse/BoldTests.cs:
```diff
@@ -11,8 +11,8 @@
+using Microsoft.Toolkit.Parsers.Markdown.Blocks;
+using Microsoft.Toolkit.Parsers.Markdown.Inlines;
 
 namespace UnitTests.Markdown.Parse
```
Should there be a new bold+italic test suite?

michael-hawker replied:
As mentioned we can resolve in #1798.

michael-hawker commented on UnitTests/UnitTests.csproj:
```diff
@@ -192,6 +192,10 @@
     <Content Include="Assets\Wide310x150Logo.scale-200.png" />
   </ItemGroup>
   <ItemGroup>
```
We should add a code block test since its new.  Anything else new?

WilliamABradley replied:
The original Code Block test still works.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Core/ExtensionMethods.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <summary>
+        /// Regular expression of HTML tags to remove.
+        /// </summary>
+        private static readonly Regex RemoveHtmlTagsRegex = new Regex(@"(?></?\w+)(?>(?:[^>'""]+|'[^']*'|""[^""]*"")*)>");
```
Will do

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Enums/ListStyle.cs:
```diff
@@ -10,26 +10,21 @@
 // THE CODE OR THE USE OR OTHER DEALINGS IN THE CODE.
 // ******************************************************************
 
-namespace Microsoft.Toolkit.Uwp.Services.Rss
```
Git got a little confused.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Enums/MarkdownBlockType.cs:
```diff
@@ -10,58 +10,61 @@
-using System;
-
-namespace Microsoft.Toolkit.Uwp.Services.Rss
+namespace Microsoft.Toolkit.Parsers.Markdown.Enums
```
yep

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Inlines/BoldItalicTextInline.cs:
```diff
@@ -49,14 +51,14 @@ internal static void AddTripChars(List<Common.InlineTripCharHelper> tripCharHelp
+            if (markdown == null || markdown.Length < 6 || start + 6 >= maxEnd)
             {
                 return null;
             }
```
This should handle it, if not I'll add it to my fix for #1798.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Microsoft.Toolkit.Parsers.csproj:
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  
+  <PropertyGroup>
```
I updated this csproj with the required info.

WilliamABradley commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/CodeRenderer.cs:
```diff
@@ -0,0 +1,160 @@
+            }
+
+            _codeView = GetTemplateChild("codeView") as RichTextBlock;
+            _printButton = GetTemplateChild("PrintButton") as Button;
```
Hey, we ourselves have a PrintHelper service, I'm not sure this PR relates to ethics :)

michael-hawker replied:
Hehe... no worries.  It can be a larger debate for another day.  Not to worry then.

WilliamABradley commented on Microsoft.Toolkit.Uwp.SampleApp/Controls/CodeRenderer/CodeRenderer.cs:
```diff
@@ -1,270 +0,0 @@
-﻿// ******************************************************************
```
How do you mean?

WilliamABradley replied:
Hassle > Value

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Blocks/ListBlock.cs:
```diff
@@ -389,6 +398,7 @@ public override string ToString()
                     case ListStyle.Bulleted:
                         result.Append("* ");
```
It was a pre-existing implementation.

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Inlines/StrikethroughTextInline.cs:
```diff
@@ -71,21 +73,21 @@ internal static Common.InlineParseResult Parse(string markdown, int start, int m
             var result = new StrikethroughTextInline();
             result.Inlines = Common.ParseInlineChildren(markdown, innerStart, innerEnd);
-            return new Common.InlineParseResult(result, start, innerEnd + 2);
+            return new InlineParseResult(result, start, innerEnd + 2);
```
Yep

WilliamABradley commented on Microsoft.Toolkit.Parsers/Markdown/Inlines/LinkAnchorInline.cs:
```diff
@@ -0,0 +1,131 @@
+
+            // Find the end of the span.  The end sequence ('-->')
+            var innerStart = start + 2;
+            int innerEnd = Common.IndexOf(markdown, "</a>", innerStart, maxEnd);
```
Not anymore :) It now uses `OrdinalIgnoreCase`

WilliamABradley commented:
@michael-hawker Made all changes against review.

michael-hawker commented on Microsoft.Toolkit/Extensions/StringExtensions.cs:
```diff
@@ -21,6 +22,26 @@ public static class StringExtensions
+        /// <summary>
+        /// Regular expression for removing comments.
+        /// </summary>
+        private static readonly Regex RemoveCommentsRegex = new Regex("<!--.*?-->", RegexOptions.Singleline);
```
Should preface these as Html comments/scripts/styles, eh?

michael-hawker commented on Microsoft.Toolkit/Extensions/StringExtensions.cs:
```diff
@@ -30,5 +51,91 @@ public static bool IsEmail(this string str)
+        }
+
+        /// <summary>
+        /// Truncates the specified string to the specified length.
```
Should mention it trims too.

WilliamABradley commented:
@michael-hawker I performed the requested changes.
